dird-conf: rewrite schedule parser

florian-at-bareos opened the pull request:
Fixes #2137: fix schedule parsing

This PR rewrites the schedule together with the way it's parsed and printed.

Also changed the following behaviour:
- if no run schedule is specified, default to "at 00:00" (i.e. run every day at 00:00) instead of currently "hourly" (i.e. run every full hour)
- warn the user if "daily", "weekly" or "monthly" keywords are used. The do nothing and are just a source of confusion
- warn the user if the specified schedule doesn't run in the next 366 days

This PR fixes the following bugs:
- week of month not being calculated correctly
- minutes not being taken into account if multiple time-of-day are specified ("at 20:15, at 20:10" is currently interpreted as  "at 20:10, at 20:10")

and fixes the documentation for Schedule -> Run so that it actually matches the implementation


The biggest advantage of having this new schedule parser is that we can unit test it, both the parsing and the schedule execution.

### Thank you for contributing to the Bareos Project!

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [ ] Is the PR title usable as CHANGELOG entry?
- [ ] Purpose of the PR is understood
- [ ] Commit descriptions are understandable and well formatted
- [ ] Required backport PRs have been created
- [ ] Correct milestone is set

##### Source code quality
- [ ] Source code changes are understandable
- [ ] Variable and function names are meaningful
- [ ] Code comments are correct (logically and spelling)
- [ ] Required documentation changes are present and part of the PR

##### Tests
- [ ] Decision taken that a test is required (if not, then remove this paragraph)
- [ ] The choice of the type of test (unit test or systemtest) is reasonable
- [ ] Testname matches exactly what is being tested
- [ ] On a fail, output of the test leads quickly to the origin of the fault

florian-at-bareos committed:
scheduler: add several time enums

together with their respective literals

florian-at-bareos committed:
scheduler: add schedule class

florian-at-bareos committed:
scheduler: add Deformat method for Schedule

florian-at-bareos committed:
run: add Schedule member to RunResource

florian-at-bareos committed:
scheduler: add modulo support

florian-at-bareos committed:
scheduler: add hourly support

florian-at-bareos committed:
scheduler: replace date_time_mask for config dump

florian-at-bareos committed:
scheduler: rework Schedule class

such that it integrates with the current scheduling implementation

florian-at-bareos committed:
scheduler: replace date_time_mask with schedule

Note that this commit changes behaviour:
"show_scheduled_preview" will now show the schedule for the next 60
minutes instead of the current hour, e.g. at 16:50 will now show
16:50-17:50 instead of 16:50-17:00.

florian-at-bareos committed:
refactor: remove unconsitent constant usage

At some point, it probably makes sense to introduce time constants but
across the entire codebase and not only one file.

florian-at-bareos committed:
refactor: remove unused DateTimeMask

as well as the RunResource::minute member

florian-at-bareos committed:
run_conf: remove outdated code

florian-at-bareos committed:
schedule: fix bug in Contains(Mask) method

florian-at-bareos committed:
schedule: improve schedule behaviour

1. if no schedule is specified, default to "daily at 00:00" instead of
"hourly"
2. deprecate "daily", "weekly" and "monthly" keywords since they do
nothing and are just confusing
3. warn if "hourly" and specific times like "at 20:00" are defined

florian-at-bareos committed:
scheduler: fix TimeOfDay with am and pm

florian-at-bareos committed:
scheduler: add hourly minute specification

to integrate with existing implementatoin

florian-at-bareos committed:
docs: update Schedule -> Run documentation

florian-at-bareos committed:
refactor: apply bareos-check-sources

and moved some unused methods to private

sebsura commented on core/src/dird/run_conf.cc:
```diff
@@ -34,11 +34,189 @@
 #include "lib/keyword_table_s.h"
```
```
Schedule {
  Name = Test
  Run = Daily at 18
}
```
this leads to the error
```
Config error: Could not parse Run directive because of illegal token "at"
            : line 16, col 20 of file etc/bareos/bareos-dir.d/director/bareos-dir.conf
  Run = Daily at 18
```
which is a bit misleading.


By the way:  On current master the situation is much worse.  The message i get is this:
```
bareos-dir: dird/check_catalog.cc:72-0 Could not open Catalog "MyCatalog", database "regress_restore": cats/postgresql.cc:235 Unable to connect to PostgreSQL server. Database=regress_restore User=ssura
Possible causes: SQL server not running; password incorrect; max_connections exceeded.
(Verbindung ist ein NULL-Zeiger
)

bareos-dir ERROR TERMINATION
Please correct the configuration in etc/bareos/bareos-dir.d/*/*.conf
```
so clearly something went really wrong during parsing :)

florian-at-bareos replied:
> Generally, there are way too few sanity checks here. A parsers job is not only to produce "machine readable data", but to also validate the input. This seems to be completely missing here.
> 
> A simple typo
> 
> ```
> Schedule {
>   Name = "Even Weeks"
>   Run = w02/w02 Friat 23:10
> }
> ```
> 
> which clearly produces nonsense is parsed "correctly" without any warnings:
> 
> ```
> * show schedules
> Schedule {
>   Name = "Even Weeks"
>   Run = w02/w02 fri w03 at 00:00
> }
> ```
> 
> But this is clearly not what the user intended!

That's very bad! This code should generally detect errors but as you also mentioned i need to be more exact on my `FromString` methods.
Here `friat` is parsed as Friday because only the first 3 letters need to match and 23:10 matches with WeekOfYear because we skip the '2' thinking it assumes it's a 'w' and then parse the int `3:100` as 3 xD

sebsura reviewed (changes requested):
I did not look at everything yet, but its clear to me that this still needs much more work.

The parser needs to actually check what its parsing and error out if its not right instead of producing a bad config!

sebsura commented on core/src/dird/run_conf.cc:
```diff
@@ -34,11 +34,189 @@
+template <class T, std::enable_if_t<std::is_same_v<T, int>, int> = 0>
+std::optional<int> FromString(const std::string& str)
+{
+  try {
+    return std::stoi(str);
+  } catch (...) {
+    return std::nullopt;
+  }
```
We should also probably capture the error information here somehow.  This may be useful to the user when they need to fix it!

florian-at-bareos commented on core/src/dird/run_conf.cc:
```diff
@@ -34,11 +34,189 @@
+              First& first,
+              Rest&... rest);
+
+// FromString
+// :: int
```
The issue is that I also want to have partial specialization for `Range<T>` and so.
Partial specialization for methods is forbidden so we have to do it like that for templates like `Range<T>`.

Now if we want your approach for non templates like `int` we get an error because of ambiguity because now the `Range<T>` method is a potential candidate but also your generic
```
template <typename T> std::optional<T> FromString(const std::string&);
```
method.

Therefore no method can be a specialization and we need enable_if.

sebsura replied:
You cannot have partial specialisations of functions, but nothing stops you from creating structs/classes instead.
I.e. instead of
```
std::optional parsed = FromString<int>(int_string);
```
you can just say
```
std::optional parsed = Parser<int>::parse(int_string);
```
see [here](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:main.cpp,fontScale:14,fontUsePx:'0',j:6,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:8,positionColumn:1,positionLineNumber:8,selectionStartColumn:1,selectionStartLineNumber:8,startColumn:1,startLineNumber:8),source:'%23include+%3Coptional%3E%0A%23include+%3Cstring%3E%0A%0Atemplate+%3Ctypename+T%3E+struct+Parser%3B%0Atemplate+%3C%3E+struct+Parser%3Cint%3E%0A%7B%0A++++static+std::optional%3Cint%3E+parse(const+std::string%26+s)+%7B+return+0%3B+%7D%0A%7D%3B%0A%0Atemplate+%3Ctypename+T%3E%0Astruct+range+%7B+%7D%3B%0A%0Atemplate+%3Ctypename+T%3E+struct+Parser%3Crange%3CT%3E%3E%0A%7B%0A++++static+std::optional%3Crange%3CT%3E%3E+parse(const+std::string%26+s)+%7B+return+%7B%7D%3B+%7D%0A%7D%3B'),l:'5',n:'0',o:main.cpp,t:'0')),header:(),k:33.333333333333336,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g142,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:6),l:'5',n:'0',o:'+x86-64+gcc+14.2+(Editor+%236)',t:'0')),header:(),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+14.2',editorid:6,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+14.2+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

sebsura replied:
Another alternative is to use function overloading instead of template specialisation.

florian-at-bareos replied:
I like
```
template<class > struct Parser {};
```

florian-at-bareos commented on core/src/dird/date_time.cc:
```diff
@@ -53,30 +52,54 @@ static int LastDayOfMonth(int year, int month)
+  day_of_week = tm.tm_wday;
+  hour = tm.tm_hour;
+  minute = tm.tm_min;
+  second = tm.tm_sec;
+  dst_ = tm.tm_isdst;
+  gmt_offset_ = tm.tm_gmtoff;
+  time_zone_ = tm.tm_zone;
```
Yes. Since these are public can change and then the `time` member is outdated. I think it makes more sense to calculate the time dynamically from the individual components.

sebsura replied:
Should they be publically writable though ? I have not checked the code, but if we do not actually currently write to them, then maybe we should turn them into private fields instead.

florian-at-bareos replied:
Yes they should be public. At some parts of the code we change the hours or minutes and then retrieve the time.

florian-at-bareos commented on core/src/dird/dird_conf.cc:
```diff
@@ -1353,70 +1353,98 @@ static void PrintConfigRunscript(OutputFormatterResource& send,
+// :: TimeOfDay
+std::string ToString(TimeOfDay time)
+{
+  std::string hour = std::to_string(int(time.hour));
+  std::string minute = std::to_string(int(time.minute));
+  return "at " + (hour.length() == 1 ? "0" + hour : hour) + ":"
+         + (minute.length() == 1 ? "0" + minute : minute);
```
we don't, was my first thought as well :) but I think importing fmt to the director for that small line is a bit overkill.

florian-at-bareos commented on core/src/dird/schedule.h:
```diff
@@ -0,0 +1,147 @@
+template <> constexpr int kMaxValue<DayOfWeek> = 6;
+
+// Range
+template <class T> struct Range {
+  using Type = T;
+
+  bool Contains(T value) const
+  {
+    if (from <= to) {
+      return from <= value && value <= to;
+    } else {
+      return from <= value || value <= to;
+    }
+  }
+
+  T from, to;
+};
```
Agree! What do you think about the name `Interval<T>` and name its members `first` and `last` to indicate that `last` is also part of the range?

sebsura replied:
Sounds good to me.

florian-at-bareos commented on core/src/dird/date_time.h:
```diff
@@ -24,23 +24,184 @@
+    "jul", "aug", "sep", "oct", "nov", "dec",
+};
+// WeekOfYear
+enum class WeekOfYear
```
Generally I think it's more neat to have enums to ditch having this weird `value` member but in that specific case I also see that having 54 enum values is also a lot of boilerplate just to store an int from 0-53.

sebsura commented on core/src/dird/dird_conf.cc:
```diff
@@ -1353,70 +1353,98 @@ static void PrintConfigRunscript(OutputFormatterResource& send,
+template <class T> std::string ToString(const std::vector<T>& vec)
+{
+  std::string result;
+  for (size_t i = 0; i < vec.size(); ++i) {
+    result += ToString(vec.at(i));
+    if (i + 1 < vec.size()) { result += ' '; }
+  }
```
Seems like my second comment got lost.  You technically dont need `first`.  You can just do
```
if (!result.empty()) { result += ' '; }
```

florian-at-bareos commented on core/src/dird/date_time.h:
```diff
@@ -24,23 +24,184 @@
+constexpr std::array<std::string_view, 6> kWeekOfMonthLiterals
+    = {"first", "second", "third", "fourth", "fifth", "last"};
+// DayOfMonth
+enum class DayOfMonth
```
Here I think it's a little different because storing just an `int` here is a little weird because of the `last` keyword which would have `value = 5`.

sebsura replied:
See my other comment.  You dont need to have an int member.  This could just as well  be
```
struct WeekOfMonth { std::string_view name; }

WeekOfMonth selectable_months[] = { {"First"}, {"Second"}, ... };
```

florian-at-bareos commented on core/src/dird/date_time.h:
```diff
@@ -24,23 +24,184 @@
+  k30,
+};
+// DayOfWeek
+enum class DayOfWeek
+{
+  kSun,
+  kMon,
+  kTue,
+  kWed,
+  kThu,
+  kFri,
+  kSat,
+};
+constexpr std::array<std::string_view, 7> kDayOfWeekLiterals = {
+    "sun", "mon", "tue", "wed", "thu", "fri", "sat",
+};
```
I see your point! I think both approaches have up and downsides.

With your approach, we can't be sure that `DayOfWeek` is actually valid because in theory we can just assign anything to `parsed_name` so `DayOfWeek` doesn't necessarily represent a day of week but can also be in an invalid state.

sebsura replied:
Thats what private constructors are for :)  I didnt write all the details here, i just wanted to give a short overview of another approach.

florian-at-bareos commented on core/src/dird/schedule.h:
```diff
@@ -0,0 +1,147 @@
+// :: Contains
+template <class T> bool Contains(const Mask<T>& mask, T value)
+{
+  if (auto* range = std::get_if<Range<T>>(&mask)) {
+    return range->Contains(value);
+  } else if (auto* modulo = std::get_if<Modulo<T>>(&mask)) {
+    return modulo->Contains(value);
+  } else {
+    return std::get<T>(mask) == value;
+  }
```
Nice!

florian-at-bareos commented on core/src/dird/schedule.h:
```diff
@@ -0,0 +1,147 @@
+  bool TriggersOnDay(DateTime date_time) const;
+  std::vector<time_t> GetMatchingTimes(time_t from, time_t to) const;
+
+  std::vector<std::variant<Mask<MonthOfYear>,
+                           Mask<WeekOfYear>,
+                           Mask<WeekOfMonth>,
+                           Mask<DayOfMonth>,
+                           Mask<DayOfWeek>>>
+      day_masks;
```
I see! This would make things more explicit but we would also every time we do something with theses masks make sure we handle every vector and don't forget one.

florian-at-bareos replied:
If we were to separate these vectors, we would have to duplicate code like this
```
  template <class T> bool IsRestricted() const
  {
    for (const auto& mask : day_masks) {
      if (std::holds_alternative<Mask<T>>(mask)) { return true; }
    }
    return false;
  }

  template <class T> bool TriggersOn(T value) const
  {
    if (!IsRestricted<T>()) { return true; }
    for (const auto& mask : day_masks) {
      if (std::holds_alternative<Mask<T>>(mask)) {
        if (Contains(std::get<Mask<T>>(mask), value)) { return true; }
      }
    }
    return false;
  }
```
5 times

and for parsing we would have to go from
```
if (auto x = Parser<std::variant<...>>::Parse(str)) {
  ...
}
```
to
```
if (auto x = Parser<Mask<...>>::Parse(str)) {

}
else if (auto x = Parser<Mask<...>>::Parse(str)) {

}
...
```
with the risk of forgetting to check for one Mask accidentally.

What do you think?

sebsura replied:
If you define `day_masks` as
```
std::tuple<std::vector<Mask<MonthOfYear>>,
           std::vector<Mask<WeekOfYear>>,
           std::vector<Mask<WeekOfMonth>>,
           std::vector<Mask<DayOfMonth>>,
           std::vector<Mask<DayOfWeek>>>>
      day_masks;
```
-- which looks pretty ugly to be fair! -- this can be done quite easily, see here: https://godbolt.org/z/3hE7T4Ghr

sebsura replied:
If you do not like the tuple approach, then this can also be achieved manually:
```C++
std::vector<Mask<...>> months_of_year;
....

template <typename T>
std::vector<Mask<T>>& get_mask_storage() = delete;

template <>
std::vector<Mask<MonthOfYear>>& get_mask_storage<MonthOfYear>() {
  return months_of_year;
}
```

florian-at-bareos replied:
> If you define `day_masks` as
> 
> ```
> std::tuple<std::vector<Mask<MonthOfYear>>,
>            std::vector<Mask<WeekOfYear>>,
>            std::vector<Mask<WeekOfMonth>>,
>            std::vector<Mask<DayOfMonth>>,
>            std::vector<Mask<DayOfWeek>>>>
>       day_masks;
> ```
> 
> -- which looks pretty ugly to be fair! -- this can be done quite easily, see here: https://godbolt.org/z/3hE7T4Ghr

I like this!

florian-at-bareos commented on core/src/dird/dird_conf.cc:
```diff
@@ -1353,70 +1353,98 @@ static void PrintConfigRunscript(OutputFormatterResource& send,
+  std::string result = "hourly";
+  for (int minute : hourly.minutes) {
+    std::string minute_str = std::to_string(int(minute));
+    result += " at 00:"
```
Yes it is now required. I will test whether currently `hourly at :15` works but if it does do you think I should adapt this? It doesnt appear in the documentation and is probably just an artifact of the parsing or is this common usage?

florian-at-bareos replied:
it was working before as you said

florian-at-bareos commented on core/src/dird/dird_conf.cc:
```diff
@@ -1353,70 +1353,98 @@ static void PrintConfigRunscript(OutputFormatterResource& send,
+  return times.front();
+}
+
+// ToString
+// :: int
```
I like to do this to group parts of the code together when for example we have several specializations of the same method or so. But I can also remove it if it's confusing to others.

florian-at-bareos commented on core/src/dird/run_conf.cc:
```diff
@@ -34,11 +34,189 @@
+template <class T, std::enable_if_t<std::is_same_v<T, WeekOfYear>, int> = 0>
+std::optional<T> FromString(const std::string& str)
+{
+  auto index = FromString<int>(str.substr(1));
```
I used `std::string` because of `std::string_view` because `std::stoi` can't use `std::string_view` but you're right!
And I can also just use `std::from_chars` instead.

florian-at-bareos committed:
conf: make FromString methods properly validating

also now dump week and month literals capitalized

florian-at-bareos committed:
dird: rename Range<T> to Interval<T>

and rename members from, to -> first, last

florian-at-bareos committed:
dird: rename Modulo<T> members

from left, right -> remainder, divisor

florian-at-bareos committed:
dird: rename Deformat method to Scan

florian-at-bareos commented on core/src/dird/run_conf.cc:
```diff
@@ -34,11 +34,189 @@
 
+// Deformat
+template <class First, class... Rest>
+bool Deformat(std::string_view input,
+              std::string_view fmt,
+              First& first,
+              Rest&... rest);
```
Yes, i guess Scan is the right name here

florian-at-bareos committed:
dird: replace FromString methods with Parser

in order to avoid confusing std::enable_if statements.

Also replace passing by 'const std::string&' to 'std::string_view'.

florian-at-bareos committed:
config: output only first 3 digits for DayOfWeek

and fix unit test expected results

florian-at-bareos committed:
date_time: store time_zone as string

and not as string_view

florian-at-bareos committed:
run_conf: add TimeOfDay parsing sanity checks

florian-at-bareos committed:
dird_conf: update some ToString methods

- add ',' between vector elements
- address review suggestions

florian-at-bareos committed:
run_conf: get rid of Scan method

florian-at-bareos committed:
date_time: replace enums with structs

for MonthOfYear, WeekOfYear, WeekOfMonth, DayOfMonth and DayOfWeek.

florian-at-bareos committed:
schedule: improve Contains method

florian-at-bareos committed:
dird: remove unused code and unnecessary comments

florian-at-bareos commented on core/src/dird/run_conf.cc:
```diff
@@ -34,11 +34,189 @@
+  try {
+    return DayOfMonth(std::stoi(str));
+  } catch (...) {
+    return std::nullopt;
```
i changed this to use std::from_chars such that we don't need to catch exceptions here

florian-at-bareos committed:
run_conf: allow to parse TimeOfDay without hours

such that schedules like "hourly at :15" stay valid

florian-at-bareos committed:
dird: add Parser<Schedule> struct

such that we can properly write unit tests for schedule parsing

florian-at-bareos committed:
schedule: replace Schedule::day_masks type

from std::vector<std::variant<Mask<...>>>
to std::tuple<std::vector<Mask<...>>>

florian-at-bareos committed:
docs: add schedule description

florian-at-bareos committed:
systemtests: update config dump reference file

make weekdays uppercase

florian-at-bareos committed:
tests: add schedule parsing unit-test

sebsura commented on core/src/dird/scheduler_private.cc:
```diff
@@ -45,8 +45,6 @@ namespace directordaemon {
 using std::chrono::seconds;
 
 static constexpr int local_debuglevel = 200;
-static constexpr auto seconds_per_hour = seconds(3600);
-static constexpr auto seconds_per_minute = seconds(60);
```
why did we remove these ?  I would rather have descriptive names than `60`s everywhere.

Arguably these should have been defined as
```C++
seconds_per_hour = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::hours(1)).count();
...
```

sebsura commented on docs/manuals/source/manually_added_config_directive_descriptions/dir-schedule-Run.rst.inc:
```diff
@@ -36,40 +36,92 @@ Accurate=yes|no
+
+If you do not specify any of these, the schedule triggers every day.
+
+For a schedule to be triggered on a certain day, it must match with a specifier from all categories (if given):
```
```suggestion
For a schedule to be triggered on a certain day, it must match with a specifier from all specified categories:
```

sebsura reviewed (changes requested):
I wrote some comments already, but currently i cannot test this as this does not compile.

sebsura commented on core/src/tests/scheduler.cc:
```diff
@@ -284,6 +284,114 @@ enum
+      DateTime date_time{time};
+      for (ValidatorBase* validator : validators) {
+        EXPECT_TRUE((*validator)(date_time)) << "In schedule \"" << schedule << "\", matching time " << time << " did not suffice all validators:" << std::endl << validator->string() << std::endl << date_time;
+      }  
```
There are also some other formatting errors, so you probably need to run the formatter on this.

florian-at-bareos committed:
scheduler: remove constexpr in method with throws

also fix some typos

florian-at-bareos committed:
date_time: add constructors from string_view

florian-at-bareos committed:
date_time: fix tm.tm_zone error

sebsura commented on core/src/dird/date_time.cc:
```diff
@@ -91,7 +91,15 @@ time_t DateTime::GetTime() const
+  }
+  else {
+    static_assert("Unexpected type of tm.tm_zone.");
+  }
```
If you know something is const, but the type system doesnt say so (because its an old api), then you can simply use `const_cast`, i.e.

```
tm.tm_zone = const_cast<decltype(tm.tm_zone)>(time_zone_.c_str());
```

florian-at-bareos committed:
date_time: fix tm struct issues

The issue is that the struct tm has no strictly defined but only states
necessary members. So in order to account for all scenarios with
potential additional members, we store a tm struct in DateTime.

florian-at-bareos committed:
run_conf: fix optional construction issue

Now this should compile on jenkins/ubuntu.

florian-at-bareos committed:
apply bareos-check-sources

florian-at-bareos committed:
date_time: fix bug in WeekOfMonth parsing

florian-at-bareos committed:
tests: test if schedules trigger correctly

florian-at-bareos committed:
date_time: unspecify daylight-saving-time

florian-at-bareos committed:
tests: fix trigger_correctly schedule test

such that it behaves the same across different time zones

florian-at-bareos committed:
docs: fix typo

sebsura reviewed (changes requested):
The keyword "on" was removed.  You need to readd it.

florian-at-bareos committed:
schedule: readd "on" token

it does nothing

florian-at-bareos committed:
docs: fix build errors

florian-at-bareos committed:
schedule: replace time numbers with constants

florian-at-bareos committed:
schedule: fix DayOfMonth ToString bug

also add some unit test cases for schedule parsing and schedule triggers

sebsura commented on docs/manuals/source/Configuration/Director.rst:
```diff
@@ -154,102 +154,64 @@ According to the NIST (US National Institute of Standards and Technology), 12am
+      Run = mon at 20:00 # run every monday at 20:00
    }
 
-The modulo scheduler makes it easy to specify schedules like odd or even days/weeks, or more generally every n days or weeks. It is called modulo scheduler because it uses the modulo to determine if the schedule must be run or not. The second variable behind the slash lets you determine in which cycle of days/weeks a job should be run. The first part determines on which day/week the job should be ...
-w03/w05.
```
I dont see any new documentation explaining the modulo scheduler, so we should probably keep this.

sebsura commented on docs/manuals/source/manually_added_config_directive_descriptions/dir-schedule-Run.rst.inc:
```diff
@@ -36,40 +36,92 @@ Accurate=yes|no
-   <day-spec>             ::= <day> | <wday-keyword> | <day> | <wday-range> | <week-keyword> <wday-keyword> | <week-keyword> <wday-range> | daily
-   <month-spec>           ::= <month-keyword> | <month-range> | monthly
-   <date-time-spec>       ::= <month-spec> <day-spec> <time-spec>
+if no schedule is specified, we fallback to "at 00:00" (i.e. run every day at 00:00).
```
```suggestion
if no schedule is specified, we fall back to "at 00:00" (i.e. run every day at 00:00).
```
`fallback` is a noun iirc.

sebsura reviewed (changes requested):
There are some problems left still:
This is not accepted even though it should be
```
Run = 1/2 at 23:10
```

In general some of the example schedules (!) from the documentation do not work.  I suggest you create a systemtest that only has a director, with a bunch of schedules, including any current example schedules and all the old ones that are not part of the documentation any more and just check if the director starts and whether the schedules are correct;  e.g. by outputting a `status schedule` and diffing that to the given configuration.

Also feel free to include the examples that i gave into the configuration file.

I can also just crash the director via `w01/w00` (this was already a bug in master).  Could you check if its easy to fix and include that here ?
Edit:  I guess if we want to backport that fix, maybe its better to create a separate pr for it.

sebsura commented:
And also there are some file conflicts between this and master that still need to be resolved.

sebsura commented:
`w20/w05` is also accept now even though it was not before.  As `w20/w05` currently never triggers, we should probably still disallow it.  It would be great if we could have a unit test for this as well.

sebsura reviewed (changes requested):
```
bareos-dir: dird/run_conf.cc:523-0 Config warning: Run directive schedule never runs in the next 366 days
            : line 3, col 29 of file /home/ssura/Projekte/bareos/build/systemtests/tests/restore/etc/bareos/bareos-dir.d/job/backup-bareos-fd.conf
  Run = w05/w05 Mon at 08:00
```

This emits a warning that it never triggers, but its supposed to trigger according to master (once every 5 weeks).

florian-at-bareos committed:
fix "fallback" typos

florian-at-bareos committed:
schedule: fix Modulo for DayOfMonth

florian-at-bareos committed:
test: add modulo unit tests

also test parsing for invalid schedules

florian-at-bareos committed:
docs: readd modulo explanation

florian-at-bareos committed:
scheduler: extend DateTime struct

Also, make time_t dynamic so that it changes if DateTime members change.
